src/platforms/virtual/graphics/graphics.cpp

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <mir/graphics/platform.h>
+#include <optional>
+
+#include "platform.h"
+
+#include <mir/assert_module_entry_point.h>
+#include <mir/libname.h>
+#include <mir/options/program_option.h>
+
+namespace mg = mir::graphics;
+namespace mo = mir::options;
+namespace mgv = mir::graphics::virt;
+
+namespace
+{
+mir::ModuleProperties const description = {
+    "mir:virtual",
+    MIR_VERSION_MAJOR,
+    MIR_VERSION_MINOR,
+    MIR_VERSION_MICRO,
+    mir::libname()
+};
+char const* virtual_displays_option_name{"virtual-output"};
+}
+
+mir::UniqueModulePtr<mg::DisplayPlatform> create_display_platform(
+    mg::SupportedDevice const&,
+    std::shared_ptr<mo::Option> const& options,
+    std::shared_ptr<mir::EmergencyCleanupRegistry> const&,
+    std::shared_ptr<mir::ConsoleServices> const& /*console*/,
+    std::shared_ptr<mg::DisplayReport> const& report)
+{
+    mir::assert_entry_point_signature<mg::CreateDisplayPlatform>(&create_display_platform);
+
+    auto outputs = options->get<std::vector<std::string>>(virtual_displays_option_name);
+    auto output_sizes = mgv::Platform::parse_output_sizes(outputs);
+    return mir::make_module_ptr<mgv::Platform>(report, std::move(output_sizes));
+}
+
+void add_graphics_platform_options(boost::program_options::options_description& config)
+{
+    mir::assert_entry_point_signature<mg::AddPlatformOptions>(&add_graphics_platform_options);
+    config.add_options()
+        (virtual_displays_option_name,
+         boost::program_options::value<std::vector<std::string>>()
+            ->default_value(std::vector<std::string>{"1280x1024"}, "1280x1024")
+            ->multitoken(),
```

Comment:
We should not be setting a default: the user needs to **explicitly** request the virtual platform.

We likely also need options to specify them as "disconnected" (or that they should be "unused" on startup).

Reply:
> We likely also need options to specify them as "disconnected" (or that they should be "unused" on startup).

Can't be disconnected, can it, as how we'd connect them again? Unused? If you've enabled it through `--virtual-output`, why would you not use it? You can always have it disabled in your display layouts.

Reply:
I was thinking about the "dynamically add a virtual output to serve to a tablet" scenario. But we agreed this could land without making that simple.

Reply:
> I was thinking about the "dynamically add a virtual output to serve to a tablet" scenario.

Right, but to close that loop we'd need some means of connecting the disconnected outputs, which IMO is out of scope here. In fact, I think for that use case we should be creating / destroying virtual outputs, when we get to it.

Reply:
The user still has to explicitly request the virtual platform (via `--platform-display-libs=mir:virtual`). When that is specified, we give them a default virtual output.

Are you saying that the existence of a `--virtual-output` option should drive whether nor not we support a virtual platform instead?

Reply:
> The user still has to explicitly request the virtual platform (via `--platform-display-libs=mir:virtual`). When that is specified, we give them a default virtual output.

Right, but then they have to know to also add `gbm-kms`, or whatever else platform is appropriate.

> Are you saying that the existence of a `--virtual-output` option should drive whether nor not we support a virtual platform instead?

Yes, same like `--wayland-host` is handled:

https://github.com/MirServer/mir/blob/7aa25d21141872b2edaf391f68cdc46590873681/src/platforms/wayland/platform_symbols.cpp#L60-L72